src/librustc_trans/back/msvc/mod.rs

```diff
@@ -42,110 +44,133 @@ mod platform {
-        //
-        // For MSVC 14 (VS 2015) we look for the Win10 SDK and failing that we look
-        // for the Win8.1 SDK. We also look for the Universal CRT.
-        //
-        // For MSVC 12 (VS 2013) we look for the Win8.1 SDK.
-        //
-        // For MSVC 11 (VS 2012) we look for the Win8 SDK.
-        //
-        // For all other versions the user has to execute the appropriate vcvars bat
-        // file themselves to configure the environment.
-        //
-        // If despite our best efforts we are still unable to find MSVC then we just
-        // blindly call `link.exe` and hope for the best.
-        return env::var_os("VCINSTALLDIR").and_then(|dir| {
-            debug!("Environment already configured by user. Assuming it works.");
-            let mut p = PathBuf::from(dir);
-            p.push("bin");
-            p.push(binsub);
-            p.push("link.exe");
-            if !p.is_file() { return None }
-            Some(Command::new(p))
-        }).or_else(|| {
-            get_vc_dir().and_then(|(ver, vcdir)| {
-                debug!("Found VC installation directory {:?}", vcdir);
-                let linker = vcdir.clone().join("bin").join(binsub).join("link.exe");
-                if !linker.is_file() { return None }
-                let mut cmd = Command::new(linker);
-                add_lib(&mut cmd, &vcdir.join("lib").join(vclibsub));
-                if ver == "14.0" {
-                    if let Some(dir) = get_ucrt_dir() {
-                        debug!("Found Universal CRT {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("ucrt").join(libsub));
-                    }
-                    if let Some(dir) = get_sdk10_dir() {
-                        debug!("Found Win10 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    } else if let Some(dir) = get_sdk81_dir() {
-                        debug!("Found Win8.1 SDK {:?}", dir);
-                        add_lib(&mut cmd, &dir.join("um").join(libsub));
-                    }
-                } else if ver == "12.0" {
-                    if let Some(dir) = get_sdk81_dir() {
+    // For MSVC 14 we need to find the Universal CRT as well as either the
+    // Windows 10 SDK or Windows 8.1 SDK.
+    fn find_msvc_14(arch: &str) -> Option<(Command, PathBuf)> {
+        get_vc_dir("14.0").and_then(|path| {
+            get_linker(&path, arch).into_iter().zip(lib_subdir(arch)).next()
+        }).and_then(|((mut cmd, host), sub)| {
+            get_ucrt_dir().map(|dir| {
+                debug!("Found Universal CRT {:?}", dir);
+                add_lib(&mut cmd, &dir.join("ucrt").join(sub));
+            }).and_then(|_| {
+                get_sdk10_dir().map(|dir| {
+                    debug!("Found Win10 SDK {:?}", dir);
+                    add_lib(&mut cmd, &dir.join("um").join(sub));
+                }).or_else(|| {
+                    get_sdk81_dir().map(|dir| {
```

Comment:
The rightward-drift in this function may be indicative that it'd benefit from an `Option`-like `try!` macro